### PR TITLE
Fix various races in push context initialization

### PR DESCRIPTION
**Pull request by howardjohn**
This contains 3 fixes for 3 races mentioned in detail in https://github.com/istio/istio/issues/29830

1. concurrent PushContext initialization may lead to storing resulting push contexts out of order, dropping a later push context. This is fixed by simply dropping `go` on AdsPushAll. It doesn't need to be a goroutine anyways, its not doing any blocking work.

2. A race between initProxy and addCon; we get some PC-1 when we init proxy, then PC-2 starts to push, but since we are not registered, we miss the push and never get an update. The fix here is to add the connection first, but block pushes until the proxy has completed initialized

3. HasSynced returns true before push context initialization is complete, leading to early connecting proxies not getting the full state of the world. This is fixed by keeping track of the configs we have send to configupdate after we are synced (it is synchronous, and so we know if we are synced its full up to date) and the configs we have finished push context initialization of, and comparing before marking as ready.

I have run this 10k+ without failures, previously it fails every ~100 runs or so

This is built on https://github.com/istio/istio/pull/29869, will be slightly smaller when that merges

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
For 2. Isn't that envoy will send initial xds request in order when it first connects to xds server?If so, there will be no missing. Can you elaborate a little bit more about this.

**Comment by hzxuzhonghu** [comment]
For 3. I am also confused, I understand the orders is correct, but before the push happens the push context is reinitiated. How is it different for proxies whether they are old or new?

File: pilot/pkg/xds/ads.go [hunk 1]
```diff
@@ -266,6 +271,13 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream discovery.AggregatedD
 	reqChannel := make(chan *discovery.DiscoveryRequest, 1)
 	go s.receive(con, reqChannel, &receiveError)
 
+	// Wait for the proxy to be fully initialized before we start serving traffic. Because
+	// initialization doesn't have dependencies that will block, there is no need to add any timeout
+	// here. Prior to this explicit wait, we were implicitly waiting by receive() not sending to
+	// reqChannel and the connection not being enqueued for pushes to pushChannel until the
+	// initialization is complete.
+	<-con.initialized
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/ads.go:279]: This way it looks good.

**Comment by howardjohn** [comment]
> For 2. Isn't that envoy will send initial xds request in order when it first connects to xds server?If so, there will be no missing. Can you elaborate a little bit more about this.

When envoy sends the initial request,we set up the proxy struct. This has two parts - initialize the proxy data (most important is Sidecar scope) then once that is done we register the connection so it gets new pushes.

The race happens if a push is triggered between those two events. We will construct the sidecar scope with pc A, then a new pc B starts to push but is not sent to our new proxy since it isn't registered yet. Therefore it misses update B

File: pilot/pkg/xds/discovery.go [hunk 2]
```diff
@@ -86,11 +86,19 @@ type DiscoveryServer struct {
...
 	mutex sync.RWMutex
 
+	// InboundUpdates describes the number of configuration updates the discovery server has received
+	InboundUpdates *atomic.Int64
+	// CommittedUpdates describes the number of configuration updates the discovery server has
+	// received, process, and stored in the push context. If this number is less than InboundUpdates,
+	// there are updates we have not yet processed.
+	// Note: This does not mean that all proxies have received these configurations; it is strictly
+	// the push context, which means that the next push to a proxy will receive this configuration.
+	CommittedUpdates *atomic.Int64
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/xds/discovery.go:100]: The merge is already handled - otherwise all the tests would fail (at startup we are merging 100s of events)

**Comment by howardjohn** [comment]
> For 3. I am also confused, I understand the orders is correct, but before the push happens the push context is reinitiated. How is it different for proxies whether they are old or new?

This is an Istiod startup issue. When istiod starts up it should not serve any xds until it's fully synced,otherwise we will send invalid config. The problem is our syncing only ensures the controllers have synced, but doesn't ensure we have actually initialized the push context with those latest data. So a proxy that connects early, before push context has been initialized for all the initial sync updates, will be missing some services. If you look in the issue I posted an example of artificially injecting a sleep which makes this trivial to verify

**Review by costinm — APPROVED** [review]
Thanks for the detailed comments in the code !

**Pull request merged**